### PR TITLE
feat: Resume logging epochs

### PR DESCRIPTION
**Pull request by mapmeld**
## Description

As part of allowing a different x-axis on our charts, we would want an option for experiments to be logged by epoch.

If an experiment has `records_per_epoch` in its configuration, record epoch based on the batches number. The calculation reverses this existing code which had converted epochs to batches.

```python
return max((epochs * self.records_per_epoch) // self.global_batch_size, 1)
```

If epoch is not in the config, set epoch = 0 (the database already defaults to total_epochs=0 on all steps)

## Test Plan

- Run `det e create gan/gan_mnist_pl/const.yaml gan/gan_mnist_pl` and find the **trial_id**. In the database, look up `select distinct(total_epochs) from steps where trial_id = 125;` and there should be values 0-4.
- Run `det e create decision_trees/gbt_titanic_estimator/const.yaml decision_trees/gbt_titanic_estimator` and find this **trial_id**. In the database, look up `select distinct(total_epochs) from steps where trial_id = 126;` and it should only be 0.

## Checklist
- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/core/_train.py [hunk 1]
```diff
@@ -72,6 +72,7 @@ def report_training_metrics(
         steps_completed: int,
         metrics: Dict[str, Any],
         batch_metrics: Optional[List[Dict[str, Any]]] = None,
+        epoch: Optional[int] = None,
```
Comment (stoksc) [inline harness/determined/core/_train.py:75]: any change to this will need @rb-determined-ai 's eyes, since it is core api

Comment (stoksc) [reply harness/determined/core/_train.py:75]: iirc we had some sort of plan for this, and that is why it is `steps_completed` not `batches_completed`

File: harness/determined/layers/_workload_sequencer.py [hunk 2]
```diff
@@ -200,12 +201,20 @@ def train(self, num_batches: int, op: core.SearcherOperation) -> WorkloadGenerat
         metrics = response.get("metrics", {}).get("avg_metrics", {})
         batch_metrics = response.get("metrics", {}).get("batch_metrics", [])
 
+        if self.records_per_epoch > 0:
+            epoch = math.ceil(
+                self.state.steps_completed * self.global_batch_size / self.records_per_epoch
+            )
+        else:
+            epoch = 0
+
```
Comment (rb-determined-ai) [inline harness/determined/layers/_workload_sequencer.py:210]: Calculating epoch from `records_per_epoch` is a bad idea, and it will frequently be wrong.  For instance, real-life datasets aren't always the same length every epoch, and every batch does not contain `global_batch_size` records.

We added `records_per_epoch` once upon a time when the workload sequencer existed only in the master, and there was no way to access the real dataset to figure out what epoch we were on.  It was messy but it was the best we could do at the time.

Now that we have moved the wlsq into the python code, new epoch-related features should be developed using actual epoch counts, rather than depending on calculated epoch derived from `records_per_epoch`.

File: harness/determined/core/_train.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/determined/core/_train.py:75]: This is basically an additional x axis, and I'd much rather we start reporting it as a metric, rather than special-casing it and increasing the data model complexity.

Ultimately, W&B lets you select arbitrary X axes in their UI, and we also want to get there.  I think we should move in that direction.  I also think that adding `epoch` as a parameter here is effectively moving away from that direction.

File: harness/determined/layers/_workload_sequencer.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/determined/layers/_workload_sequencer.py:210]: (so you would have to report epoch as a metric from each of EstimatorTrial, TFKerasTrial, and PyTorchTrial)

**Comment by mapmeld** [comment]
ok, closing. epochs is going to come back totally separated from regular batch counts, and reported in the metrics instead of a new field
